frame/election-provider-multi-phase/src/unsigned.rs

```diff
@@ -60,24 +65,62 @@ impl From<FeasibilityError> for MinerError {
 	}
 }
 
+/// Save a given call into OCW storage.
+fn save_solution<T: Config>(call: &Call<T>) {
+	let storage = StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL);
+	storage.set(&call);
+}
+
+/// Get a saved solution from OCW storage if it exists.
+fn restore_solution<T: Config>() -> Option<Call<T>> {
+	StorageValueRef::persistent(&OFFCHAIN_CACHED_CALL).get().flatten()
+}
+
 impl<T: Config> Pallet<T> {
-	/// Mine a new solution, and submit it back to the chain as an unsigned transaction.
-	pub fn mine_check_and_submit() -> Result<(), MinerError> {
+	/// Attempt to restore a solution from cache. Otherwise, compute it fresh. Either way, submit.
+	pub fn restore_or_compute_then_submit() -> Result<(), MinerError> {
+		let call = match restore_solution() {
+			Some(call) => call,
+			None => {
+				let call = Self::mine_call()?;
+				save_solution(&call);
+				call
+			}
+		};
+		Self::submit_call(call)
+	}
+
+	/// Mine a new solution, cache it, and submit it back to the chain as an unsigned transaction.
+	pub fn mine_check_save_submit() -> Result<(), MinerError> {
+		let call = Self::mine_call()?;
+		save_solution(&call);
+		Self::submit_call(call)
+	}
+
+	/// Mine a new solution as a call. Performs all checks.
+	fn mine_call() -> Result<Call<T>, MinerError> {
```

Comment:
```suggestion
	fn mine_checked_call() -> Result<Call<T>, MinerError> {
```
(I am very keen on making it clear that any call that we mine in this code MUST PASS ALL CHECKS, meaning that it is guaranteed to be feasible and good, if submitted to the chain as an unsigned tx. As you can see, an incorrect unsigned tx actually panics, so we should avoid it as much as possible. Essentially, we are telling the validators: Hey, come use our OCW code, we promise it won't produce invalid blocks.)

Reply:
Agree. b3e49500c